File: 8-Puzzle/solution.py
import math
import numpy as np
from queue import PriorityQueue

#the opposite direction of the free tile
directions_dictionary = ['right', 'left', 'down', 'up']
visited = []

class State:
    def __init__(self, tiles, free_tile_location = None, g=0):
        self.tiles = tiles

        self.g = g
        self.h = 0
        self.children = PriorityQueue()
        self.parent = None
        self.free_tile_location = self.find_free_tile() if free_tile_location is None else free_tile_location

    def __eq__(self, other):
        return np.allclose(self.tiles, other.tiles)

    def __lt__(self, other):
        return self.f < other.f

    def __le__(self, other):
        return self.f <= other.f

    def __gt__(self, other):
        return self.f > other.f

    def __ge__(self, other):
        return self.f >= other.f

    @property
    def f(self):
        return self.g + self.h

    def to_array(self):
        return [col for row in self.tiles for col in row]

    def to_matrix(self):
        return self.tiles

    def find_free_tile(self):
        size = len(self.tiles)
        for i in range(size):
            for j in range(size):
                if self.tiles[i][j] == 0:
                    return (i, j)

    def generate_children(self, free_tile_target):
        x, y = self.free_tile_location
        #direction of free tile
        directions = [[x, y - 1], [x, y + 1], [x - 1, y], [x + 1, y]]
        size = len(self.tiles)
        for i, dir in enumerate(directions):
            if dir[0] < 0 or dir[0] > size - 1 or dir[1] < 0 or dir[1] > size - 1:
                continue
            self.tiles[x][y], self.tiles[dir[0]][dir[1]] = self.tiles[dir[0]][dir[1]], self.tiles[x][y]
            state = State(np.copy(self.tiles), dir, self.g + 1)
            state.compute_heuristics_manhattan(free_tile_target[0] * size + free_tile_target[1])
            state.parent = (self, directions_dictionary[i])
            self.children.put(state)
            self.tiles[x][y], self.tiles[dir[0]][dir[1]] = self.tiles[dir[0]][dir[1]], self.tiles[x][y]

    def compute_heuristics_manhattan(self, free_tile_index):
        state = self.tiles
        size = len(state)
        heuristics = 0
        for i in range(size):
            for j in range(size):
                num = state[i][j]
                if num != 0:
                    if num < free_tile_index:
                        num -= 1

                    num_real_row = num // size
                    num_real_col = num % size
                else:
                    num_real_row = free_tile_index // size
                    num_real_col = free_tile_index % size
                heuristics += abs(i - num_real_row) + abs(j - num_real_col)

        self.h = heuristics

def has_solution(arr_from_tiles):
    inversions_count = 0
    for i in range(0, len(arr_from_tiles)):
        for j in range(i + 1, len(arr_from_tiles)):
            if arr_from_tiles[i] and arr_from_tiles[j] and arr_from_tiles[i] > arr_from_tiles[j]:
                inversions_count += 1

    return inversions_count % 2 == 0

def get_final_state(tiles, free_tile_index):
    res_array = []
    offset = 1
    for i in range(0, len(tiles)**2) :
        if i == free_tile_index:
            res_array.append(0)
            offset -= 1
        else:
            res_array.append(i + offset)

    res = np.array(res_array).reshape(size, size)
    return res

def a_star(current, goal, limit):
    if current == goal:
        return current

    if limit == 0:
        return False

    visited.append(current)
    current.generate_children(goal.free_tile_location)
    while not current.children.empty():
        child = current.children.get()
        if child not in visited:
            #print(child.tiles)
            #print(limit)
            res = a_star(child, goal, limit - 1)
            if res:
                return res

    return False


def print_result(steps, final_state):
    print(steps)
    res = []
    while final_state.parent:
        res.append(final_state.parent[1])
        final_state = final_state.parent[0]
    res.reverse()
    print("\n".join(res))

def iterative_deepening_a_star(start, goal, limit):
    for l in range(1, limit + 1):
        print("L: ", l)
        visited.clear()
        res = a_star(start, goal, l)
        if res:
            print_result(l, res)
            return

n = int(input())
size = int(math.sqrt(n + 1))
if size * size != n + 1:
    raise AssertionError("Wrong size!")

free_tile_index = int(input())
free_tile_index = free_tile_index if free_tile_index != -1 else n
tiles = []
for r in range(size):
    row = list(map(lambda x: int(x), input().split(' ')))
    tiles.append(row)

start_state = State(tiles)

if not has_solution(start_state.to_array()):
    print("Puzzle not solvable!")
    exit(0)

start_state.compute_heuristics_manhattan(free_tile_index)
final_state = State(get_final_state(tiles, free_tile_index), (free_tile_index // size, free_tile_index % size))
print("FINAL STATE")
print(final_state.tiles)
iterative_deepening_a_star(start_state, final_state, 50)

File: 8-Puzzle/solution.cpp
#include <cmath>
#include <vector>
#include <algorithm>
#include <iostream>
#include <string.h>
#include <chrono>
#include <iomanip>
using namespace std;
using namespace std::chrono;

int directions[4][2] = { {-1, 0}, {1, 0}, {0, -1}, {0, 1} };
string directionsDictionary[4] = { "down", "up", "right", "left" };

int n, freeTileIndex, puzzleSize;

struct Table
{
    vector<vector<int>> table;
    int h = 0;
    int x, y;
    string moveTo;

    bool operator==(const Table& t) const
    {
        for (int i = 0; i < puzzleSize; i++)
            for (int j = 0; j < puzzleSize; j++)
                if (table[i][j] != t.table[i][j])
                    return false;
        return true;
    }

    void calculateHeuristic()
    {
        int realRow, realCol;
        this->h = 0;
        for (int i = 0; i < puzzleSize; i++)
            for (int j = 0; j < puzzleSize; j++) {
                int num = table[i][j];
                if (num != 0) {
                    if (num < freeTileIndex) {
                        num--;
                    }

                    realRow = num / puzzleSize;
                    realCol = num % puzzleSize;
                }
                else {
                    realRow = freeTileIndex / puzzleSize;
                    realCol = freeTileIndex % puzzleSize;
                }

                this->h += abs(i - realRow) + abs(j - realCol);
            }
    }
};

int calculateHeuristicForPosition(const int& row, const int& column, int num)
{
    int realRow, realCol;
    if (num != 0) {
        if (num < freeTileIndex) {
            num--;
        }

        realRow = num / puzzleSize;
        realCol = num % puzzleSize;
    }
    else {
        realRow = freeTileIndex / puzzleSize;
        realCol = freeTileIndex % puzzleSize;
    }
    return abs(row - realRow) + abs(column - realCol);
}

Table startState, finalState;

bool hasSolution(const Table& t) {
    int inversions = 0, index = 0;
    vector<int> flattened;
    for (int i = 0; i < puzzleSize; i++)
        for (int j = 0; j < puzzleSize; j++)
            flattened.push_back(t.table[i][j]);

    for (int i = 0; i < puzzleSize; i++)
        for (int j = i + 1; j < puzzleSize; j++)
            if (flattened[i] && flattened[j] && flattened[i] > flattened[j])
                inversions++;

    return inversions % 2 == 0;
}

void input()
{
    scanf_s("%d", &n);
    scanf_s("%d", &freeTileIndex);
    puzzleSize = (int)sqrt(n + 1);
    freeTileIndex = freeTileIndex == -1 ? n : freeTileIndex;

    for (int i = 0; i < puzzleSize; i++) {
        vector<int> row;
        for (int j = 0; j < puzzleSize; j++)
        {
            int num;
            scanf_s("%d", &num);
            row.push_back(num);
            if (num == 0)
            {
                startState.x = i;
                startState.y = j;
            }
        }
        startState.table.push_back(row);
    }

    int cnt = 1;
    int freeTileX = freeTileIndex / puzzleSize, freeTileY = freeTileIndex % puzzleSize;
    for (int i = 0; i < puzzleSize; i++) {
        vector<int> row;
        for (int j = 0; j < puzzleSize; j++) {
            int num = (i == freeTileX && j == freeTileY) ? 0 : cnt++;
            row.push_back(num);
        }
        finalState.table.push_back(row);
    }

    finalState.h = 0;
    finalState.x = freeTileIndex / puzzleSize;
    finalState.y = freeTileIndex % puzzleSize;
    if (!hasSolution(finalState)) {
        cout << "No solution!" << endl;
        exit(1);
    }
    startState.calculateHeuristic();
}

int aStar(vector<Table>& path, int g, int bound)
{
    Table current = path.back();
    int f = g + current.h;
    if (f > bound) return f;
    if (current == finalState) return 0;
    int min = INT_MAX;

    //generate children
    Table child = current;
    int t;
    for (int i = 0; i < 4; i++) {
        int newX = current.x + directions[i][0];
        int newY = current.y + directions[i][1];

        if (newX < 0 || newX >= puzzleSize ||
            newY < 0 || newY >= puzzleSize)
            continue;

        child.h = current.h -
            calculateHeuristicForPosition(newX, newY, current.table[newX][newY]) +
            calculateHeuristicForPosition(current.x, current.y, current.table[newX][newY]) -
            calculateHeuristicForPosition(current.x, current.y, current.table[current.x][current.y]) +
            calculateHeuristicForPosition(newX, newY, current.table[current.x][current.y]);

        swap(child.table[current.x][current.y], child.table[newX][newY]);
        if (std::find(path.begin(), path.end(), child) != path.end()) {
            swap(child.table[current.x][current.y], child.table[newX][newY]);
            continue;
        }
        child.x = newX;
        child.y = newY;
        child.moveTo = directionsDictionary[i];

        path.push_back(child);
        t = aStar(path, g + 1, bound);
        if (t == 0) return 0;
        if (t < min) min = t;
        swap(child.table[current.x][current.y], child.table[newX][newY]);
        path.pop_back();
    }
    return min;
}

void printPath(vector<Table>& path) {
    cout << path.size() - 1 << endl;
    for (int i = 1; i < path.size(); i++) {
        cout << path[i].moveTo << endl;
    }
}

void iterativeDeepeningAStar() {
    int bound = startState.h;
    vector<Table> path;
    path.push_back(startState);
    int t;
    auto start = high_resolution_clock::now();
    while (true) {
        t = aStar(path, 0, bound);
        if (t == 0) break;
        bound = t;
    }
    auto stop = high_resolution_clock::now();
    printPath(path);
    auto duration = duration_cast<milliseconds>(stop - start);
    cout << fixed;
    cout << setprecision(2);
    cout << "Time taken: " << (float)duration.count() / 1000 << "s" << endl;
}

int main()
{
    input();
    iterativeDeepeningAStar();

    return 0;
}

/**
Some tests:

8
-1
1 2 3
4 5 6
0 7 8

8
-1
1 0 3
4 2 5
7 8 6

8
-1
4 0 3
5 2 1
7 8 6

8
-1
0 8 6
5 4 3
2 1 7

15
-1
8 11 1 12
5 0 15 9
13 14 4 10
7 3 6 2
**/
